### PR TITLE
nixos/security/wrappers: make well-typed

### PR DESCRIPTION
**Pull request by rnhmjoj**
###### Motivation for this change

The security.wrappers option is morally a set of submodules but it's
actually (un)typed as a generic attribute set. This is bad for several
reasons:

1. Some of the "submodule" option are not document;
2. the default values are not documented and are chosen based on
   somewhat bizarre rules (issue #23217);
3. It's not possible to override an existing wrapper due to the
   dumb types.attrs.merge strategy;
4. It's easy to make mistakes that will go unnoticed, which is
   really bad given the sensitivity of this module (issue #47839).

This makes the option a proper set of submodule and add strict types and
descriptions to every sub-option. Considering it's not yet clear if the
way the default values are picked is intended, ~this reproduces the current
behavior, but it's now documented explicitly~. I gave up and switched to sane
fixed defaults.

###### Things done

- [x] Tested via `sudo doas firejail` NixOS tests
- [x] Decided what to do about the default owner, group
- [x] Compared the permission before/after change
- [x] Document changes
- [x] Update any module relying on the bizzare default rules

@dotlambda @peterhoeg @ixmatus

### CODE REVIEW

**Comment by rnhmjoj** [comment]
I figured a trick to simplify the testing: instead of checking the permissions
of the actual binaries, which requires downloading the packages and building a
VM, I compiled a list of all modules using `security.wrappers`, enabled all of
them and and diffed the resulting activation script of the system.

Here's the list:
```
{ ... }:

{
  security.duosec.ssh.enable = true;
  security.duosec.secretKeyFile = "/dev/null";
  security.duosec.integrationKey = "";
  security.duosec.host = "";

  security.doas.enable = true;
  security.sudo.enable = true;
  security.pam.usb.enable = true;
  security.polkit.enable = true;
  security.chromiumSuidSandbox.enable = true;

  programs.atop.enable = true;
  programs.noisetorch.enable = true;
  programs.ccache.enable = true;
  programs.wshowkeys.enable = true;
  programs.slock.enable = true;
  programs.wireshark.enable = true;
  programs.iftop.enable = true;
  programs.udevil.enable = true;
  programs.iotop.enable = true;
  programs.mtr.enable = true;
  programs.mosh.enable = true;
  programs.kbdlight.enable = true;
  programs.liboping.enable = true;
  programs.bandwhich.enable = true;
  programs.traceroute.enable = true;
  programs.x2goserver.enable = true;
  programs.wavemon.enable = true;
  programs.singularity.enable = true;
  programs.firejail.enable = true;

  programs.captive-browser.enable = true;
  programs.captive-browser.interface = "";

  services.locate.enable = true;
  services.physlock.enable = true;
  services.fcron.enable = true;
  services.atd.enable = true;
  services.cron.enable = true;
  services.xserver.desktopManager.cde.enable = true;
  services.xserver.desktopManager.plasma5.enable = true;
  services.xserver.desktopManager.gnome.enable = true;
  services.xserver.desktopManager.enlightenment.enable = true;
  services.weechat.enable = true;

  services.mame.enable = true;
  services.mame.user = "mame";
  services.mame.emuAddr = "0.0.0.0";
  services.mame.hostAddr = "0.0.0.0";

  services.opensmtpd.enable = true;
  services.opensmtpd.serverConfiguration = "";

  services.gnome.gnome-keyring.enable = true;
  services.postfix.enable = true;
  services.exim.enable = true;
  services.kbfs.enable = true;
  services.dbus.enable = true;
  services.smokeping.enable = true;
  services.netdata.enable = true;
  services.incron.enable = true;

  services.zabbixProxy.enable = true;
  services.zabbixProxy.server = "0.0.0.0";
  services.zabbixServer.enable = true;

  virtualisation.spiceUSBRedirection.enable = true;
  virtualisation.libvirtd.enable = true;
  virtualisation.virtualbox.host.enable = true;

  boot.supportedFilesystems = [ "ecryptfs" ];

  fileSystems."/".device = "/dev/sda";
  boot.loader.grub.device = "/dev/sda";
  networking.hostName = "a";
  networking.domain = "b";
}
```
and the script can be found by running `nix repl nixos/ -I nixos-config=test.nix`.

Now I'm sure the wrappers settings of all the modules in Nixpkgs are preserved.

**Comment by peterhoeg** [comment]
How do you feel about pulling this related PR in together with this?

https://github.com/NixOS/nixpkgs/pull/117188

File: nixos/modules/programs/ccache.nix
Comment (peterhoeg) [inline nixos/modules/programs/ccache.nix:31]: We should have a dedicated user - not a shared "nobody" user.

File: nixos/modules/security/pam.nix
Comment (peterhoeg) [inline nixos/modules/security/pam.nix:864]: Similarly here with `nogroup`.

File: nixos/modules/services/mail/postfix.nix
Comment (peterhoeg) [inline nixos/modules/services/mail/postfix.nix:676]: Same thing here with all of these.

File: nixos/modules/programs/ccache.nix
Comment (rnhmjoj) [reply nixos/modules/programs/ccache.nix:31]: I'm trying to preserve the current permissions/ownership in this PR. I agree it's better to avoid nobody, but I don't want risking breaking program or modules.

**Comment by rnhmjoj** [comment]
> How do you feel about pulling this related PR in together with this?

Uhm, I don't want to break compatibility too much: I'm already changing the default ownership which could cause trouble for modules outside Nixpkgs.

Checking only files under /nix/store should be safe enough, though.

File: nixos/modules/security/wrappers/default.nix
Comment (rnhmjoj) [inline nixos/modules/security/wrappers/default.nix:252]: This seems to be the best way to attach a build time check, inspired by the udev module.

File: nixos/modules/security/wrappers/default.nix
Comment (symphorien) [inline nixos/modules/security/wrappers/default.nix:46]: default and example should probably use `=` instead of `+`, otherwise the actual mode depends on the umask of the activation script

File: nixos/modules/security/wrappers/default.nix
Comment (symphorien) [inline nixos/modules/security/wrappers/default.nix:41]: I know I suggested that, but root is a good default for the group only for setuid, not for setgid, so maybe it's a better option to force the user to be always explicit and not have a default.

File: nixos/modules/services/mail/opensmtpd.nix
Comment (symphorien) [inline nixos/modules/services/mail/opensmtpd.nix:106]: cc @obadz @Ekleog what owner should /run/wrappers/bin/smtpctl have?

File: nixos/modules/services/mail/postfix.nix
Comment (symphorien) [reply nixos/modules/services/mail/postfix.nix:676]: cc @globin @dotlambda @lewo postfix maintainers: what owner should /run/wrappers/bin/sendmail have? As well as the other wrappers below

File: nixos/modules/programs/ccache.nix
Comment (symphorien) [reply nixos/modules/programs/ccache.nix:31]: cc @kira-bruneau @r-burns as maintainers of ccache

File: nixos/modules/security/pam.nix
Comment (symphorien) [reply nixos/modules/security/pam.nix:864]: this one is setuid root so we can make it owned by root:root I think.

**Review by symphorien — COMMENTED** [review]
this misses release notes.

File: nixos/modules/programs/ccache.nix
Comment (kira-bruneau) [reply nixos/modules/programs/ccache.nix:31]: Yep, I think it makes sense to add a dedicated user. @r-burns and I didn't originally write this (we made ourselves maintainers after noticing it was abandoned), but it looks like as long as the group stays as `nixbld`, there shouldn't be any problems with permissions.

I'll try it out the change to be sure.

**Comment by kira-bruneau** [comment]
Two of my other NixOS modules `gamemode` & `replay-sorcery` also use the security wrappers, is there anything I'll need to consider/change there?

File: nixos/modules/security/wrappers/default.nix
Comment (rnhmjoj) [reply nixos/modules/security/wrappers/default.nix:46]: Makes sense, I'm applying this.

File: nixos/modules/security/wrappers/default.nix
Comment (rnhmjoj) [reply nixos/modules/security/wrappers/default.nix:41]: Right, I'll take a while to change this: most NixOS modules needs to be changed.

File: nixos/modules/programs/ccache.nix
Comment (kira-bruneau) [reply nixos/modules/programs/ccache.nix:31]: It looks like volth originally wrote the module, but I just relalized they don't have a github account anymore :disappointed:

Comment (rnhmjoj) [reply nixos/modules/programs/ccache.nix:31]: They still have an account on the nixos.org forum, if you want to try.

File: nixos/modules/security/wrappers/default.nix
Comment (rnhmjoj) [reply nixos/modules/security/wrappers/default.nix:41]: At this point I wonder if it's better to make everything explicit and removing also the `setuid = true;` default.

**Comment by rnhmjoj** [comment]
> Two of my other NixOS modules gamemode & replay-sorcery also use security wrappers, is there anything I'll need to consider/change there?

Not sure: these are `setcap` wrappers and the original `security.wrapper` logic made these `root:root`.
So, they're not defaulting to `nobody:nogroup`, but I'm not sure that `root:root` is what you need.

**Comment by kira-bruneau** [comment]
> Not sure: these are `setcap` wrappers and the original `security.wrapper` logic made these `root:root`.
> So, they're not defaulting to `nobody:nogroup`

Ok that makes sense, thanks!

> but I'm not sure that `root:root` is what you need.

I thought `root:root` is required for capabilities? I could be wrong though. Thoses process will still start as your current user & group since the setuid & setgid option's won't be applied.

**Comment by rnhmjoj** [comment]
> I thought root:root is required for capabilities?

I don't know: I checked `capabilities(7)`, but it doesn't mention ownership in relation to file capabilities. I guess we'll have to try.

**Comment by rnhmjoj** [comment]
So, updates:

- I've rebased the changes on top of master
- completely removed the defaults (forcing module authors to think about the wrapper permissions)
- used `=` instead of `+` in chmod
- converted `nogroup` in setuid wrappers to `root`

I've checked again that the wrappers in all modules in Nixpkgs haven't changed (besides the aformentioned changes) with the method described at the top of this thread.

File: nixos/modules/security/pam.nix
Comment (rnhmjoj) [reply nixos/modules/security/pam.nix]: Done

**Comment by symphorien** [comment]
Looks good to me with release notes.
As a final proposition about the default for owner and group (but current state of the PR is fine by me, feel free to ignore): we can set it to null, and replace null by root except for the combination or (setuid, owner) and (setgid, group). Those must be set explcitely. Possible implementation:
```diff
diff --git a/nixos/modules/security/pam.nix b/nixos/modules/security/pam.nix
index 8b1f653d4e2..9c8cd706677 100644
--- a/nixos/modules/security/pam.nix
+++ b/nixos/modules/security/pam.nix
@@ -871,7 +871,6 @@ in
       unix_chkpwd = {
         setuid = true;
         owner = "root";
-        group = "root";
         source = "${pkgs.pam}/sbin/unix_chkpwd.orig";
       };
     };
diff --git a/nixos/modules/security/wrappers/default.nix b/nixos/modules/security/wrappers/default.nix
index f20e36449b0..f9aa55f29fc 100644
--- a/nixos/modules/security/wrappers/default.nix
+++ b/nixos/modules/security/wrappers/default.nix
@@ -32,12 +32,14 @@ let
         '';
       };
     options.owner = lib.mkOption
-      { type = lib.types.str;
-        description = "The owner of the wrapper program.";
+      { type = lib.types.nullOr lib.types.str;
+        default = null;
+        description = "The owner of the wrapper program. When null, defaults to root for setgid and setcap wrappers. Required for setuid wrappers.";
       };
     options.group = lib.mkOption
-      { type = lib.types.str;
-        description = "The group of the wrapper program.";
+      { type = lib.types.nullOr lib.types.str;
+        default = null;
+        description = "The group of the wrapper program. When null, defaults to root for setuid and setcap wrappers. Required for setgid wrappers.";
       };
     options.permissions = lib.mkOption
       { type = fileModeType;
@@ -93,13 +95,17 @@ let
     , ...
     }:
     assert (lib.versionAtLeast (lib.getVersion config.boot.kernelPackages.kernel) "4.3");
+    let
+      owner_ = if owner == null then "root" else owner;
+      group_ = if group == null then "root" else group;
+    in
     ''
       cp ${securityWrapper}/bin/security-wrapper "$wrapperDir/${program}"
       echo -n "${source}" > "$wrapperDir/${program}.real"
 
       # Prevent races
       chmod 0000 "$wrapperDir/${program}"
-      chown ${owner}.${group} "$wrapperDir/${program}"
+      chown ${owner_}.${group_} "$wrapperDir/${program}"
 
       # Set desired capabilities on the file plus cap_setpcap so
       # the wrapper program can elevate the capabilities set on
@@ -121,13 +127,17 @@ let
     , permissions
     , ...
     }:
+    let
+      group_ = if group == null && !setgid then "root" else group;
+      owner_ = if owner == null && !setuid then "root" else owner;
+    in
     ''
       cp ${securityWrapper}/bin/security-wrapper "$wrapperDir/${program}"
       echo -n "${source}" > "$wrapperDir/${program}.real"
 
       # Prevent races
       chmod 0000 "$wrapperDir/${program}"
-      chown ${owner}.${group} "$wrapperDir/${program}"
+      chown ${owner_}.${group_} "$wrapperDir/${program}"
 
       chmod "u${if setuid then "+" else "-"}s,g${if setgid then "+" else "-"}s,${permissions}" "$wrapperDir/${program}"
     '';
@@ -185,6 +195,16 @@ in
 
   ###### implementation
   config = {
+    assertions = lib.flatten (lib.mapAttrsToList (name: opts: [
+      {
+        assertion = opts.setuid -> opts.owner != null;
+        message = "${name} is a setuid wrapper so security.wrappers.${name}.owner must be set explicitely.";
+      }
+      {
+        assertion = opts.setgid -> opts.group != null;
+        message = "${name} is a setgid wrapper so security.wrappers.${name}.group must be set explicitely.";
+      }
+    ]) config.security.wrappers);
 
     security.wrappers =
       let
```

File: nixos/modules/security/wrappers/default.nix
Comment (symphorien) [reply nixos/modules/security/wrappers/default.nix:254]: apparently runCommandNoCCLocal does not exist anymore:
```
$  nix-instantiate nixos/tests/gnome.nix
error: attribute 'runCommandNoCCLocal' missing, at /home/symphorien/src/nixpkgs/nixos/modules/security/wrappers/default.nix:254:47
(use '--show-trace' to show detailed location information)
```

**Comment by rnhmjoj** [comment]
> As a final proposition about the default for owner and group (but current state of the PR is fine by me, feel free to ignore): we can set it to null, and replace null by root except for the combination or (setuid, owner) and (setgid, group). Those must be set explcitely. Possible implementation:

I thought about doing something like this, but now I realised it's better to make all settings mandatory: the least astonishment principle applies here. It's much easier to figure what this module does by looking at
```
wrapper = 
  { setuid = true;
    owner = "root";
    group = "root";
    source = "...";
  };
```
compared to

```
wrapper = 
  { setuid = true;
    owner = "root";
    source = "...";
  };
```
or even worse `wrapper.source = "..."`.

Still, you gave me an idea: we should enforce the mutual exclusivity of `mkSetcapProgram` and `mkSetuidProgram`.

**Comment by rnhmjoj** [comment]
New update:
- fixed three more modules that had managed to elude my regex (though `services.mail.sendmailSetuidWrapper`).
- updated the examples in `security.wrappers` option
- added release notes

File: nixos/modules/security/wrappers/default.nix
Comment (rnhmjoj) [reply nixos/modules/security/wrappers/default.nix:281]: Fixed.

File: nixos/modules/security/wrappers/default.nix
Comment (rnhmjoj) [reply nixos/modules/security/wrappers/default.nix:44]: I dad to revert this. Apparently it undoes the other changes in
```
chmod "u${if setuid then "+" else "-"}s,g${if setgid then "+" else "-"    }s,${permissions}"
```
Anyway, the script does a `chmod 0000` before, so it shouldn't be a proble.

**Comment by rnhmjoj** [comment]
@GrahamcOfBorg test sudo doas firejail

**Comment by symphorien** [comment]
Do we leave all the `owner = "nobody";` then?

**Comment by rnhmjoj** [comment]
@symphorien We could wait for the module maintainers to reply, or just merge this now and open an issue/pr to address the `nobody`s later. I don't wont to keep this open for too long, though: keeping this pr up-to-date is time-consuming.

File: nixos/modules/services/x11/desktop-managers/cde.nix
Comment (rnhmjoj) [inline nixos/modules/services/x11/desktop-managers/cde.nix:53]: This module appears unnmaintained.

**Review by symphorien — APPROVED** [review]

**Comment by Izorkin** [comment]
After this update error start netdata service:
```
Checking that Nix store paths of all wrapped programs exist... FAIL
The path /nix/store/r0jjm4gc5rsi82fyvv8nfzljbizgg1ba-netdata-1.31.0/libexec/netdata/plugins.d/freeipmi.plugin.org does not exist!
Please, check the value of `security.wrappers."freeipmi.plugin".source`.

trying https://github.com/openSUSE/catatonit/archive/v0.1.6.tar.gz
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   126  100   126    0     0    414      0 --:--:-- --:--:-- --:--:--   413
100 21546    0 21546    0     0  30888      0 --:--:-- --:--:-- --:--:-- 30888
unpacking source archive /build/v0.1.6.tar.gz

builder for '/nix/store/fk0bn549zi8bbkq0z2kzq89csa45dmnm-ensure-all-wrappers-paths-exist.drv' failed with exit code 1

```

**Comment by rnhmjoj** [comment]
Strange, I can't reproduce this on master (5aef5e8d1) with `nix build -f nixos/tests/netdata.nix`.
Do you have more information?

**Comment by Izorkin** [comment]
Check with this patch:
```diff
diff --git a/nixos/tests/netdata.nix b/nixos/tests/netdata.nix
index 0f26630da9d..b4b40d1941d 100644
--- a/nixos/tests/netdata.nix
+++ b/nixos/tests/netdata.nix
@@ -10,6 +10,15 @@ import ./make-test-python.nix ({ pkgs, ...} : {
     netdata =
       { pkgs, ... }:
         {
+          nixpkgs.overlays = [
+            (self: super: {
+              netdata = super.netdata.override {
+                withIpmi = false;
+                withDebug = false;
+              };
+            })
+          ];
+
           environment.systemPackages = with pkgs; [ curl jq ];
           services.netdata.enable = true;
         };

```

**Comment by ArdaXi** [comment]
Just noticed the effect of this merge, physlock didn't get updated and now breaks because the `user` option no longer exists: https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/security/physlock.nix#L132
```error: The option `security.wrappers.physlock.user' does not exist. Definition values:
       - In `/nix/store/c2chs5a98zdw8fg9vidcwadf3vsvmwj4-source/nixos/modules/services/security/physlock.nix': "root"```

**Comment by rnhmjoj** [comment]
> Just noticed the effect of this merge, physlock didn't get updated and now breaks

Yes, I missed this, but it shows the type system is working as intended: `user` never existed. What the author meant is probably `owner`.

**Comment by rnhmjoj** [comment]
> nixpkgs.overlays = [
>   (self: super: {
>     netdata = super.netdata.override {
>       withIpmi = false;
>       withDebug = false;
>     };
>   })
> ];

The netdata module, by deault, creates a wrapper for this plugin: if you remove it, the wrapper fails to build and rightly so. Before this change the activation script would produce a broken wrapper:
```
/run/wrappers/bin/freeipmi.plugin: cannot run `/nix/store/.../netdata-1.31./libexec/netdata/plugins.d/freeipmi.plugin.org: No such file or directory
```

If you really need to do this, you have to add an option to `services.netdata` for disabling plugins, which overrides the package and disable the creation of the wrapper.

**Comment by dasJ** [comment]
This broke the `vmware-guest` module:
> error: The option `security.wrappers.vmware-user-suid-wrapper.owner' is used but not defined.

**Comment by rnhmjoj** [comment]
@dasJ where is that wrapper defined? I can't find it in Nixpkgs.

**Comment by lukegb** [comment]
FYI, this caused Hydra to stop being able to evaluate nixpkgs. Not sure _why_ yet.

**Comment by rnhmjoj** [comment]
> FYI, this caused Hydra to stop being able to evaluate nixpkgs. Not sure why yet.

ofBorg eval check passed right before merging, I don't see how.

**Comment by lukegb** [comment]
It's not an evaluation failure, per se, it's `hydra-eval-jobs` doing its usual thing of failing inside nlohmann JSON without providing any additional context.

**Comment by vcunat** [comment]
OfBorg evaluates the package set.  These JSON errors typically happen inside NixOS (manual, or modules through tests).

**Comment by rnhmjoj** [comment]
I'm going through all the errors I'm seeing in https://hydra.nixos.org/jobset/nixos/trunk-combined#tabs-errors.
I've identified and fixed 10 so far, but nothing about this PR, yet.

**Comment by vcunat** [comment]
Well, most of the errors in there are "intentional non-support" of a particular platform/configuration.

**Comment by rnhmjoj** [comment]
There are a lot of NixOS errors too: https://github.com/NixOS/nixpkgs/pull/138918

**Pull request merged**
